pkg/testrig/testrig.go

```diff
@@ -0,0 +1,221 @@
+
+	byteValue, err := ioutil.ReadAll(templateFile)
+	if err != nil {
+		return nil, err
+	}
+	var vars template.ServiceMap
+	err = json.Unmarshal(byteValue, &vars)
+	if err != nil {
+		return nil, err
+	}
+
+	return vars, nil
+}
+
+func generateAppService(serviceName string, serviceTmpl template.Service, outputDir string) map[string]interface{} {
+	build := make(map[string]interface{})
+	build["context"] = "."
+	build["dockerfile"] = filepath.Join(outputDir, serviceName, "Dockerfile")
+	block := make(map[string]interface{})
+	block["build"] = build
+	block["ports"] = []string{fmt.Sprintf("%v:%v", serviceTmpl.Port, serviceTmpl.Port)}
+	return block
+}
+
+func generateDbService(serviceName string) map[string]interface{} {
+	block := make(map[string]interface{})
+	block["image"] = "postgres:latest"
+	block["ports"] = []string{"5432:5432"}
+	// piece of magic to make Postgres execute our script on startup
+	block["volumes"] = []string{
+		fmt.Sprintf("../gen/%v/%v.sql:/docker-entrypoint-initdb.d/%v.sql",
+			serviceName, serviceName, serviceName),
+	}
+	environment := make(map[string]string)
+	environment["POSTGRES_USER"] = "someuser"
+	environment["POSTGRES_PASSWORD"] = "somepassword"
+	environment["POSTGRES_DB"] = "somedb"
+	block["environment"] = environment
+	return block
+}
+
+func generateCompose(file *os.File, composeServices map[string]interface{}) error {
+	data := make(map[string]interface{})
+	data["version"] = "3.3"
+	data["services"] = composeServices
+
+	dataRaw, err := yaml.Marshal(data)
+	if err != nil {
+		return err
+	}
+
+	_, err = file.Write(dataRaw)
+	if err != nil {
+		return err
+	}
+	err = file.Sync()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func processMustacheTemplate(file *os.File, template string, data template.Service, outputDir string) error {
+	// mustache needs generic map
+	rawData, err := json.Marshal(&data)
+	if err != nil {
+		return err
+	}
+	var genericData map[string]interface{}
+	err = json.Unmarshal(rawData, &genericData)
+	if err != nil {
+		return err
+	}
+	genericData["outputDir"] = outputDir
+
+	renderedTemplate, err := mustache.Render(template, genericData)
+	if err != nil {
+		return err
+	}
+	_, err = file.WriteString(renderedTemplate)
+	if err != nil {
+		return err
+	}
+	err = file.Sync()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func generateMain(serviceName string, service template.Service, outputDir string, needDb bool) error {
+	output, err := os.Create(filepath.Join(outputDir, serviceName, "main.go"))
+	if err != nil {
+		return err
+	}
+	defer output.Close()
+
+	mainTemplate := template.GetMainStub()
+	if needDb {
+		mainTemplate = template.GetMainDbStub()
+	}
+
+	err = processMustacheTemplate(output, mainTemplate, service, outputDir)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func generateDockerfile(serviceName string, service template.Service, outputDir string) error {
+	output, err := os.Create(filepath.Join(outputDir, serviceName, "Dockerfile"))
```

Comment:
All creation of files should be using the afero.fs within sysl, not os (this allows for using in memory fs for systems that can't use os file systems)

Reply:
oh, right, will do